File: BCKEND/TypeScript/BancoFinanceiro/src/Model/Transacao.ts
import { Cliente } from "./Cliente";
import { TipoTransacao } from "./TipoTransacao";

export class Transacao {
    private tipo: TipoTransacao;
    private valor: number;
    private data: Date;
    private clienteTransferencia?: Cliente;
    private infoPagamento?: string;
    private tipoValor: string;

    constructor(tipo: TipoTransacao, valor: number, tipoValor: string, 
        clienteTransferencia?: Cliente, infoPagamento?: string) {
        this.tipo = tipo;
        this.valor = valor;
        this.data = new Date();
        if(clienteTransferencia) this.clienteTransferencia = clienteTransferencia;
        if(infoPagamento) this.infoPagamento = infoPagamento;
        this.tipoValor = tipoValor;
    }

    public getTipo(): TipoTransacao {
        return this.tipo;
    }

    public getValor(): number {
        return this.valor;
    }

    public getData(): Date {
        return this.data;
    }

    public getClienteTransferencia(): Cliente | undefined {
        return this.clienteTransferencia;
    }

    public getInfoPagamento(): string | undefined {
        return this.infoPagamento;
    }

    public getTipoValor(): string {
        return this.tipoValor;
    }

    public toString(): string {
        let info = `${this.data.toLocaleString()} \t ${this.tipo} \t ${this.tipoValor} R$ ${this.valor.toFixed(2)}`;
        if(this.clienteTransferencia) {
            info += `\tCliente de Transferência: ${this.clienteTransferencia.getNome()}`;
        }
        if(this.infoPagamento) {
            info += `\tInformações do Pagamento: ${this.infoPagamento}`;
        }
        return info;
    }
}

File: BCKEND/JAVA/BancoFinanceiro/src/Model/Transacao.java
package Model;

import java.util.Date;

public class Transacao {
	
	private TipoTransacao tipo;
	private Date data;
	private double valor;
	private Cliente clienteTransferencia;
	private String infoPagamento;
	private char tipoValor;
	
	public Transacao(TipoTransacao tipo, double valor, char tipoValor) {
		super();
		this.tipo = tipo;
		this.data = new Date();
		this.valor = valor;
		this.clienteTransferencia = null;
		this.infoPagamento = null;
		this.tipoValor = tipoValor;
	}
	
	public Transacao(TipoTransacao tipo, double valor, Cliente clienteTransferencia, char tipoValor) {
		super();
		this.tipo = tipo;
		this.data = new Date();
		this.valor = valor;
		this.clienteTransferencia = clienteTransferencia;
		this.infoPagamento = null;
		this.tipoValor = tipoValor;
	}
	
	public Transacao(TipoTransacao tipo, double valor, String infoPagamento, char tipoValor) {
		super();
		this.tipo = tipo;
		this.data = new Date();
		this.valor = valor;
		this.clienteTransferencia = null;
		this.infoPagamento = infoPagamento;
		this.tipoValor = tipoValor;
	}

	public TipoTransacao getTipo() {
		return tipo;
	}

	public Date getData() {
		return data;
	}

	public double getValor() {
		return valor;
	}

	public Cliente getClienteTransferencia() {
		return clienteTransferencia;
	}

	public char getTipoValor() {
		return tipoValor;
	}
	
	public String getInfoPagamento() {
		return infoPagamento;
	}

	@Override
	public String toString() {
		if (clienteTransferencia != null) {
			return  data.getDate() + "/" + 	(data.getMonth()+1) + "/" + 
					(data.getYear()+1900) + "\t" + tipo + ", " + 
					clienteTransferencia.getNome() + "\t" + 
					tipoValor + " R$ " + valor;
		} else if(infoPagamento != null) {
			return  data.getDate() + "/" + 	(data.getMonth()+1) + "/" + 
					(data.getYear()+1900) + "\t" + tipo + ", " + 
					infoPagamento + "\t" + 
					tipoValor + " R$ " + valor;
		} else {
			return  data.getDate() + "/" + 	(data.getMonth()+1) + "/" + 
					(data.getYear()+1900) + "\t" + tipo + ", " +
					tipoValor + " R$ " + valor;
		}		
	}	

}
